[jit] use memory locations instead of values for working set

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21227 [jit] make analyzeConservative more conservative
* #21224 [jit] cleanup templated implementation of mayAlias
* **#21223 [jit] use memory locations instead of values for working set**
* #21222 [jit] remove multisets from work set
* #21221 [jit] cleanups to alias analysis

Topological moves now work off of memoryDAG elements. This is useful
because

1. it overall reduces the amount of work done (since it's now
capped by the number of unique memory locations in the graph)

2. we can now express reads/writes to wildcards directly (which we
couldn't do before since they are elements without a corresponding
value).

Differential Revision: [D15592317](https://our.internmc.facebook.com/intern/diff/D15592317)

suo committed:
[jit] use memory locations instead of values for working set

Topological moves now work off of memoryDAG elements. This is useful
because

1. it overall reduces the amount of work done (since it's now
capped by the number of unique memory locations in the graph)

2. we can now express reads/writes to wildcards directly (which we
couldn't do before since they are elements without a corresponding
value).

suo committed:
Update on "[jit] use memory locations instead of values for working set"

[jit] use memory locations instead of values for working set

Topological moves now work off of memoryDAG elements. This is useful
because

1. it overall reduces the amount of work done (since it's now
capped by the number of unique memory locations in the graph)

2. we can now express reads/writes to wildcards directly (which we
couldn't do before since they are elements without a corresponding
value).

gh-metadata: pytorch pytorch 21223 gh/suo/47/head

suo committed:
Update on "[jit] use memory locations instead of values for working set"

[jit] use memory locations instead of values for working set

Topological moves now work off of memoryDAG elements. This is useful
because

1. it overall reduces the amount of work done (since it's now
capped by the number of unique memory locations in the graph)

2. we can now express reads/writes to wildcards directly (which we
couldn't do before since they are elements without a corresponding
value).

gh-metadata: pytorch pytorch 21223 gh/suo/47/head

zdevito commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -80,24 +80,27 @@ bool AliasDb::hasWriters(const Value* v) const {
 
-  if (isWriteCacheStale_) {
-    rebuildWriteCache();
+  const auto& memoryLocations = elementMap_.at(v)->getMemoryLocations();
```
This double hashes elementMap_

zdevito commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -80,24 +80,27 @@ bool AliasDb::hasWriters(const Value* v) const {
-  if (isWriteCacheStale_) {
-    rebuildWriteCache();
+  const auto& memoryLocations = elementMap_.at(v)->getMemoryLocations();
+  for (const auto& pr : writeIndex_) {
```
This seems expensive, was it this way before?

suo replied:
We used to cache the whole set of written locations when comparisons were more inefficient. I can add that back.

zdevito commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -1086,10 +1050,14 @@ bool AliasDb::writesToWildcard(Node* n) const {
-  return std::any_of(writes.cbegin(), writes.cend(), [&](const Value* v) {
-    return mayAliasWildcard(v);
-  });
+  // Are any of these memoryLocs a wildcard element?
```
The `writesToWildcard` function seems suspicious given wildcards are not really "special" and just the heap. It seems like it has a single use in `hasUntrackedMutation`. Looking at DCE, it probably also needs a pass to make it understand the wildcard set as the 'heap' rather than special annotation.

suo replied:
Good point I'll stack a PR to fix

suo committed:
Update on "[jit] use memory locations instead of values for working set"

[jit] use memory locations instead of values for working set

Topological moves now work off of memoryDAG elements. This is useful
because

1. it overall reduces the amount of work done (since it's now
capped by the number of unique memory locations in the graph)

2. we can now express reads/writes to wildcards directly (which we
couldn't do before since they are elements without a corresponding
value).

gh-metadata: pytorch pytorch 21223 gh/suo/47/head

suo committed:
Update on "[jit] use memory locations instead of values for working set"

[jit] use memory locations instead of values for working set

Topological moves now work off of memoryDAG elements. This is useful
because

1. it overall reduces the amount of work done (since it's now
capped by the number of unique memory locations in the graph)

2. we can now express reads/writes to wildcards directly (which we
couldn't do before since they are elements without a corresponding
value).

gh-metadata: pytorch pytorch 21223 gh/suo/47/head
